mockkubeapiserver/postresource.go

```diff
@@ -73,8 +74,42 @@ func (req *postResource) Run(ctx context.Context, s *MockKubeAPIServer) error {
 		obj.SetGeneration(1)
 	}
 
+	if err := preCreateHook(ctx, obj); err != nil {
+		return err
+	}
+
 	if err := resource.CreateObject(ctx, id, obj); err != nil {
 		return err
 	}
 	return req.writeResponse(obj)
 }
+
+func preCreateHook(ctx context.Context, obj *unstructured.Unstructured) error {
+	gvk := obj.GroupVersionKind()
+	if gvk.Group == "" && gvk.Version == "v1" && gvk.Kind == "Secret" {
+		stringDataAny, found := obj.Object["stringData"]
+		if found {
+			stringDataMap, ok := stringDataAny.(map[string]any)
+			if !ok {
+				return fmt.Errorf("unexpected type for v1.Secret stringData, got %T", stringDataAny)
+			}
+			dataAny, dataFound := obj.Object["data"]
+			if !dataFound {
+				dataAny = make(map[string]any)
+				obj.Object["data"] = dataAny
+			}
+			dataMap, ok := dataAny.(map[string]any)
+			if !ok {
+				return fmt.Errorf("unexpected type for v1.Secret data, got %T", dataAny)
+			}
+			for k, vAny := range stringDataMap {
+				v, ok := vAny.(string)
+				if !ok {
+					return fmt.Errorf("unexpected type for v1.Secret stringData entry, got %T", vAny)
+				}
+				dataMap[k] = base64.StdEncoding.EncodeToString([]byte(v))
```

Comment:
I'm not familiar enough with what Go will do through all of these casts; will this actually end up mutating `obj`? Or does it just mutate the `dataMap` we've created, but not actually propagate those changes back?

Reply:
So the answer is yes, and the test will (hopefully!) prove it.  I'll have a go at rewriting it to be more self-evidently true though, I agree there's too much happening here.